File: python/src/tradingflow/operators/traders/benchmark.py
"""Benchmark operator - frictionless ideal portfolio replication."""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from ... import ArrayView, Handle, NodeKind, Operator


@dataclass(slots=True)
class BenchmarkState:
    # Configuration.
    num_stocks: int
    initial_cash: float
    use_adjusts: bool

    # Portfolio state.
    cash: np.ndarray = field(default_factory=lambda: np.array(0.0))
    shares: np.ndarray = field(default_factory=lambda: np.empty(0))
    last_adjust: np.ndarray = field(default_factory=lambda: np.empty(0))
    last_close: np.ndarray = field(default_factory=lambda: np.empty(0))

    # Rebalance signal observed on the previous tick, awaiting
    # execution at today's open.  `None` when there is no pending
    # signal.  Deferring execution by one tick keeps the signal's
    # information set (universe construction, factor values, ...)
    # strictly older than the execution price.
    pending_positions: np.ndarray | None = None


class Benchmark(
    Operator[
        ArrayView[np.float64],  # soft positions (num_stocks,)
        ArrayView[np.float64],  # close prices (num_stocks,)
        ArrayView[np.float64],  # adjusts (num_stocks,)
        ArrayView[np.float64],  # upper price limit (num_stocks,)
        ArrayView[np.float64],  # lower price limit (num_stocks,)
        ArrayView[np.float64],  # output: (holdings_value, cash)
        BenchmarkState,
    ]
):
    """Frictionless benchmark that replicates soft position weights exactly.

    On every tick:

    1. Adjusts held shares for dividend reinvestment via forward
       adjustment factor changes.
    2. If a rebalance signal was received on the **previous** tick,
       first force-liquidates held positions in stocks with no valid
       exec price today (suspended or delisted) at their last valid
       close, then rebalances the remaining tradable stocks to their
       target allocations via a single net trade at today's close
       (MOC-style):
       `target_shares = pending_positions * current_value / close_price`,
       executed as fractional shares with no transaction fees and no
       lot rounding.  Trades that would violate today's daily price
       limit are blocked: buys are skipped in stocks whose close is at
       or above ``upper_limit`` (limit-up, no sellers), and sells are
       skipped in stocks whose close is at or below ``lower_limit``
       (limit-down, no buyers).  Suspended or delisted stocks hold
       zero shares post-rebalance.
    3. If the soft-positions input was updated this tick, stores it as
       the pending signal for execution on the **next** tick.  Trading
       is therefore deferred by one tick, so today's signal (derived
       from data observable at today's close) is executed at tomorrow's
       close - the upstream signal construction never has access to the
       execution-price information set.
    4. Outputs a 2-element array `(holdings_value, cash)` where
       `holdings_value` is positions valued at closing prices and
       `cash` is the cash balance.  Total portfolio value is their sum.

    Parameters
    ----------
    soft_positions
        Soft position weights, shape `(num_stocks,)`.
    close
        Stacked unadjusted close prices, shape `(num_stocks,)`.  Used
        both as the execution price (MOC at today's close) and as the
        mark-to-market price.
    adjusts
        Stacked forward adjustment factors, shape `(num_stocks,)`.
    upper_limit, lower_limit
        Today's daily price-limit thresholds, shape `(num_stocks,)`.
        On every tick the rebalance blocks **buys** in stocks whose
        close has reached or exceeded ``upper_limit`` and **sells** in
        stocks whose close has reached or fallen below ``lower_limit``
        - matching A-shares limit-locked behaviour where the order book
        is one-sided at the limit price.  `NaN` in either handle is
        treated as "no constraint" for that side.  Typically computed
        upstream via
        [`build_price_limits`][common.build_price_limits].
    initial_cash
        Starting capital.
    use_adjusts
        If `True`, account for dividend reinvestment via adjustment
        factors (total return index).  If `False`, use raw prices
        (price index).

    Notes
    -----
    The rebalance cadence is controlled by upstream: the benchmark
    rebalances exactly when the soft-positions input produces.

    **NaN prices (suspended / delisted stocks).**  A stock with `NaN`
    close on the current tick is handled as follows:

    - **Valuation (every tick)**: portfolio value uses the *last valid
      close* carried forward, so held shares of a suspended stock keep
      contributing at their most recent known price instead of dropping
      out.
    - **Rebalance**: stocks with a finite, positive current close are
      rebalanced via a single net trade (delta shares from current to
      target).  Stocks with non-finite or non-positive close cannot be
      traded at the close; if we held any, the simulator force-closes
      them at their last valid close (an idealisation that is not
      achievable in live trading) so they hold zero shares
      post-rebalance.  No fresh entry is ever made into a stock whose
      current close is invalid - capital earmarked for it remains in
      cash.
    """

    def __init__(
        self,
        soft_positions: Handle,
        close: Handle,
        adjusts: Handle,
        upper_limit: Handle,
        lower_limit: Handle,
        *,
        initial_cash: float,
        use_adjusts: bool,
    ) -> None:
        assert len(soft_positions.shape) == 1, "Soft positions input must have shape (num_stocks,)."
        assert len(close.shape) == 1, "Close input must have shape (num_stocks,)."
        assert len(upper_limit.shape) == 1, "Upper limit input must have shape (num_stocks,)."
        assert len(lower_limit.shape) == 1, "Lower limit input must have shape (num_stocks,)."
        assert (
            soft_positions.shape[0] == close.shape[0] == upper_limit.shape[0] == lower_limit.shape[0]
        ), "All input handles must agree on the number of stocks."

        self._num_stocks = soft_positions.shape[0]
        self._initial_cash = initial_cash
        self._use_adjusts = use_adjusts

        super().__init__(
            inputs=(soft_positions, close, adjusts, upper_limit, lower_limit),
            kind=NodeKind.ARRAY,
            dtype=np.float64,
            shape=(2,),
            name=type(self).__name__,
        )

    def init(
        self,
        inputs: tuple[ArrayView[np.float64], ArrayView[np.float64], ArrayView[np.float64]],
        timestamp: int,
    ) -> BenchmarkState:
        n = self._num_stocks
        return BenchmarkState(
            num_stocks=n,
            initial_cash=self._initial_cash,
            use_adjusts=self._use_adjusts,
            cash=np.array(self._initial_cash),
            shares=np.zeros(n),
            last_adjust=np.ones(n),
            last_close=np.full(n, np.nan),
        )

    @staticmethod
    def compute(
        state: BenchmarkState,
        inputs: tuple[ArrayView[np.float64], ArrayView[np.float64], ArrayView[np.float64]],
        output: ArrayView[np.float64],
        timestamp: int,
        produced: tuple[bool, ...],
    ) -> bool:
        soft_positions = inputs[0].value()
        closes = inputs[1].value()
        adjusts = inputs[2].value()
        upper_limit = inputs[3].value()
        lower_limit = inputs[4].value()

        # Adjust shares for dividends (reinvesting all dividends).
        if state.use_adjusts:
            valid_adjusts = np.isfinite(adjusts) & (adjusts > 0)
            adjust_mask = valid_adjusts & (state.last_adjust > 0)
            state.shares[adjust_mask] *= adjusts[adjust_mask] / state.last_adjust[adjust_mask]
            state.last_adjust[valid_adjusts] = adjusts[valid_adjusts]

        # Update the last-valid-close carry-forward for stocks that
        # ticked this cycle; suspended stocks retain their previous
        # last-valid close.  Done BEFORE the rebalance because today's
        # close is now both the execution price and the mark-to-market
        # price - the same value is used to size, fill, and value the
        # post-trade position, so there is no intraday lookahead.
        close_tick = np.isfinite(closes)
        state.last_close[close_tick] = closes[close_tick]

        # Execute pending rebalance from the previous tick: signal
        # observed one tick ago, executed at today's close (MOC-style),
        # sized at today's close.  Trading on the closing-auction print
        # is the dominant venue for institutional systematic strategies
        # and is closer to VWAP than open execution.  The one-tick
        # delay keeps the signal information set strictly older than
        # the execution price (signal at close[t] -> execution at
        # close[t+1]), and using today's close for both sizing and
        # execution removes any intraday open/close mismatch.
        if state.pending_positions is not None:
            pending = state.pending_positions
            valid_exec = np.isfinite(closes) & (closes > 0)

            # Step 1: force-liquidate held positions in stocks with no
            # valid exec price today (suspended or delisted) at their
            # last valid close - the simulator assumes an idealised
            # exit even when no open-market trade is actually possible.
            force_liq = (state.shares != 0) & ~valid_exec & np.isfinite(state.last_close)
            state.cash += np.sum(state.shares[force_liq] * state.last_close[force_liq])
            state.shares[force_liq] = 0.0

            # Step 2: compute portfolio value post force-liquidation
            # using today's close, then rebalance tradable stocks to
            # target via a single net trade at today's close.  (In a
            # frictionless benchmark this coincides with full liquidate
            # + re-enter; the delta form keeps the logic consistent
            # with `SimpleTrader`, which must use it to avoid doubling
            # fees.)
            held = (state.shares != 0) & np.isfinite(state.last_close)
            current_value = state.cash + np.sum(state.shares[held] * state.last_close[held])
            safe_closes = np.where(valid_exec, closes, 1.0)
            target_shares = pending * current_value / safe_closes
            trade_shares = np.where(valid_exec, target_shares - state.shares, 0.0)

            # Step 3: enforce A-shares-style price-limit rules.  At
            # limit-up the order book has no sellers (buys cannot fill);
            # at limit-down it has no buyers (sells cannot fill).  Block
            # the corresponding side of any limit-locked stock by
            # zeroing its trade_shares; NaN in either limit is treated
            # as "no constraint" (e.g. first tick before history exists,
            # or markets without a limit rule).
            block_buy = np.isfinite(upper_limit) & (closes >= upper_limit) & (trade_shares > 0)
            block_sell = np.isfinite(lower_limit) & (closes <= lower_limit) & (trade_shares < 0)
            trade_shares = np.where(block_buy | block_sell, 0.0, trade_shares)

            state.cash -= np.sum(trade_shares * safe_closes)
            state.shares += trade_shares

            state.pending_positions = None

        # Store new rebalance signal for execution on the NEXT tick.
        # `.value()` already returns a fresh copy, but `.copy()` makes
        # the one-tick-delay intent explicit at the storage site.
        if produced[0]:
            state.pending_positions = soft_positions.copy()

        # Output `(holdings_value, cash)`.  Total portfolio value = sum.
        held = (state.shares != 0) & np.isfinite(state.last_close)
        holdings_value = np.sum(state.shares[held] * state.last_close[held])

        # Bankruptcy check.  A leveraged or sum-zero portfolio (e.g. a
        # long-short construction with gross > 0 and net == 0) can
        # accumulate enough adverse PnL on one leg to push total NAV
        # to zero or below; in a real account the broker margin-calls
        # and force-liquidates everything.  Emulate that here by
        # zeroing cash, shares, and any pending signal whenever NAV is
        # non-positive at the tick boundary.  The wiped-out state is
        # absorbing: on subsequent ticks `current_value` stays at 0,
        # so all rebalances trade nothing and every subsequent output
        # is `(0, 0)`.  Downstream plots are responsible for rendering
        # this as a gap (NaN) rather than a literal `log(0) = -inf`.
        if not (state.cash + holdings_value > 0):
            state.shares[:] = 0.0
            state.cash = np.array(0.0)
            state.pending_positions = None
            holdings_value = 0.0

        output.write(np.array([holdings_value, state.cash], dtype=np.float64))
        return True

File: crates/tradingflow/src/operators/traders/benchmark.rs
//! The frictionless [`Benchmark`] executor.

use super::core::{TraderInputs, TraderValues, Vp};
use crate::data::{Array, ArrayView, Instant, Layout};
use crate::graph::Segment;
use crate::ports::is_eventful;

/// Frictionless benchmark executor: replicates target weights exactly, with
/// dividend reinvestment, one-tick-delayed mark-on-close execution, idealised
/// force-liquidation of suspended holdings, A-shares price-limit blocking, and a
/// bankruptcy wipe. Output is `[2]` = `[holdings_value, cash]`; NAV is their sum.
pub struct Benchmark {
    num_stocks: usize,
    initial_cash: f64,
    use_adjusts: bool,
}

impl Benchmark {
    /// `num_stocks` is the cross-section width (= the input arrays' length);
    /// `initial_cash` seeds the book (use `1.0` for a unit-NAV index);
    /// `use_adjusts` toggles dividend reinvestment via the adjust factors.
    pub fn new(num_stocks: usize, initial_cash: f64, use_adjusts: bool) -> Self {
        Self {
            num_stocks,
            initial_cash,
            use_adjusts,
        }
    }
}

/// Runtime state for [`Benchmark`]: the book (cash + per-stock shares), the
/// dividend/close carry-forwards, the rebalance signal pending one-tick
/// execution, and the `[2]` output buffer.
pub struct BenchmarkState {
    num_stocks: usize,
    use_adjusts: bool,
    cash: f64,
    shares: Vec<f64>,
    last_adjust: Vec<f64>,
    last_close: Vec<f64>,
    pending: Option<Vec<f64>>,
    out: Array<f64, 1>,
}

impl Segment for Benchmark {
    type Inputs = TraderInputs;
    type Outputs = Vp;
    type Context = Instant;
    type State = BenchmarkState;

    fn init(self, (pos, ..): TraderValues<'_>) -> BenchmarkState {
        let n = self.num_stocks;
        assert_eq!(
            pos.layout().len(),
            n,
            "Benchmark: input length {} != num_stocks {n}",
            pos.layout().len(),
        );
        BenchmarkState {
            num_stocks: n,
            use_adjusts: self.use_adjusts,
            cash: self.initial_cash,
            shares: vec![0.0; n],
            last_adjust: vec![1.0; n],
            last_close: vec![f64::NAN; n],
            pending: None,
            out: Array::zeros([2]),
        }
    }

    fn compute<'a, 'b: 'a>(
        (pos, close, adj, up, lo): TraderValues<'a>,
        state: &'b mut BenchmarkState,
        _: &Instant,
    ) -> ArrayView<'a, f64, 1> {
        let n = state.num_stocks;
        let pos_eventful = is_eventful(pos);
        let can_exec = is_eventful(close);
        let positions = pos.to_contiguous();
        let closes = close.to_contiguous();
        let adjusts = adj.to_contiguous();
        let upper = up.to_contiguous();
        let lower = lo.to_contiguous();

        // Reinvest dividends: scale held shares by the change in adjust factor.
        if state.use_adjusts {
            for i in 0..n {
                let a = adjusts[i];
                if a.is_finite() && a > 0.0 {
                    if state.last_adjust[i] > 0.0 {
                        state.shares[i] *= a / state.last_adjust[i];
                    }
                    state.last_adjust[i] = a;
                }
            }
        }

        // Carry forward the last valid close for stocks that ticked this cycle;
        // suspended stocks retain their previous last-valid close. Done before
        // the rebalance so today's close is both the sizing and marking price.
        for i in 0..n {
            if closes[i].is_finite() {
                state.last_close[i] = closes[i];
            }
        }

        // Execute the rebalance signalled one tick ago, at today's close.
        // Only an eventful close batch can execute: with no close events there
        // is no market to trade in, and the pending signal is held.
        if can_exec && let Some(pending) = state.pending.take() {
            // Step 1: idealised force-liquidation of held stocks with no valid
            // exec price today (suspended/delisted), at their last valid close.
            for i in 0..n {
                let valid_exec = closes[i].is_finite() && closes[i] > 0.0;
                if state.shares[i] != 0.0 && !valid_exec && state.last_close[i].is_finite() {
                    state.cash += state.shares[i] * state.last_close[i];
                    state.shares[i] = 0.0;
                }
            }
            // Step 2: portfolio value after force-liquidation (last-valid close).
            let mut current_value = state.cash;
            for i in 0..n {
                if state.shares[i] != 0.0 && state.last_close[i].is_finite() {
                    current_value += state.shares[i] * state.last_close[i];
                }
            }
            // Step 3: single net trade to target at today's close, tradable
            // stocks only, with A-shares price-limit blocking.
            for i in 0..n {
                let valid_exec = closes[i].is_finite() && closes[i] > 0.0;
                if !valid_exec {
                    continue;
                }
                let close = closes[i];
                let target_shares = pending[i] * current_value / close;
                let mut trade = target_shares - state.shares[i];
                let block_buy = upper[i].is_finite() && close >= upper[i] && trade > 0.0;
                let block_sell = lower[i].is_finite() && close <= lower[i] && trade < 0.0;
                if block_buy || block_sell {
                    trade = 0.0;
                }
                state.cash -= trade * close;
                state.shares[i] += trade;
            }
        }

        // Capture a new target for execution on the NEXT eventful close.
        if pos_eventful {
            state.pending = Some(positions.to_vec());
        }

        // Output [holdings_value, cash], with a bankruptcy wipe.
        let mut holdings_value = 0.0;
        for i in 0..n {
            if state.shares[i] != 0.0 && state.last_close[i].is_finite() {
                holdings_value += state.shares[i] * state.last_close[i];
            }
        }
        // Ruin check. The negation is deliberate and NOT equivalent to
        // `<= 0.0`: net worth is `NaN` when a held price is missing, and
        // `!(NaN > 0.0)` is `true` while `NaN <= 0.0` is `false`. Rewriting
        // this comparison would let a NaN-valued book keep trading.
        #[allow(clippy::neg_cmp_op_on_partial_ord)]
        if !(state.cash + holdings_value > 0.0) {
            state.shares.iter_mut().for_each(|s| *s = 0.0);
            state.cash = 0.0;
            state.pending = None;
            holdings_value = 0.0;
        }
        let out = state.out.data_mut();
        out[0] = holdings_value;
        out[1] = state.cash;
        state.out.view()
    }

    fn reset<'a, 'b: 'a>(
        _: TraderValues<'a>,
        state: &'b mut BenchmarkState,
    ) -> ArrayView<'a, f64, 1> {
        state.out.view()
    }
}

/// Frictionless benchmark executor: replicates target weights exactly.
pub fn benchmark(num_stocks: usize, initial_cash: f64, use_adjusts: bool) -> Benchmark {
    Benchmark::new(num_stocks, initial_cash, use_adjusts)
}

#[cfg(test)]
mod tests {
    use super::super::test_util::{arr, src};
    use super::*;
    use crate::graph::pool::Pool;
    use crate::graph::typed::Builder;

    #[test]
    fn benchmark_one_tick_delay_exec_and_mark() {
        let nan = f64::NAN;
        let mut b = Builder::new();
        let (pos, posv) = src(&mut b, &[nan, nan]);
        let (close, closev) = src(&mut b, &[nan, nan]);
        let (_adj, adjv) = src(&mut b, &[1.0, 1.0]);
        let (_up, upv) = src(&mut b, &[nan, nan]);
        let (_lo, lov) = src(&mut b, &[nan, nan]);
        let out = b.segment(Benchmark::new(2, 1.0, true), (posv, closev, adjv, upv, lov));
        let mut g = b.build();
        let mut pool = Pool::new(0);

        *g.state_mut(pos) = arr(&[0.5, 0.5]);
        *g.state_mut(close) = arr(&[10.0, 20.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        assert_eq!(g.view(out).as_slice().unwrap(), &[0.0, 1.0]);

        *g.state_mut(close) = arr(&[11.0, 22.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        let o = g.view(out).as_slice().unwrap().to_vec();
        assert!((o[0] - 1.0).abs() < 1e-12, "tick2 holdings {} != 1.0", o[0]);
        assert!(o[1].abs() < 1e-12, "tick2 cash {} != 0", o[1]);

        *g.state_mut(close) = arr(&[12.0, 22.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        let o = g.view(out).as_slice().unwrap().to_vec();
        let nav = o[0] + o[1];
        let expected = 0.5 * (12.0 / 11.0) + 0.5;
        assert!(
            (nav - expected).abs() < 1e-12,
            "tick3 NAV {} != {}",
            nav,
            expected
        );
    }

    #[test]
    fn benchmark_reinvests_dividends() {
        let nan = f64::NAN;
        let mut b = Builder::new();
        let (pos, posv) = src(&mut b, &[nan]);
        let (close, closev) = src(&mut b, &[nan]);
        let (adj, adjv) = src(&mut b, &[1.0]);
        let (_up, upv) = src(&mut b, &[nan]);
        let (_lo, lov) = src(&mut b, &[nan]);
        let out = b.segment(Benchmark::new(1, 1.0, true), (posv, closev, adjv, upv, lov));
        let mut g = b.build();
        let mut pool = Pool::new(0);

        *g.state_mut(pos) = arr(&[1.0]);
        *g.state_mut(close) = arr(&[10.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        *g.state_mut(close) = arr(&[10.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        let o = g.view(out).as_slice().unwrap().to_vec();
        assert!(
            (o[0] - 1.0).abs() < 1e-12 && o[1].abs() < 1e-12,
            "fully invested: {o:?}"
        );

        *g.state_mut(adj) = arr(&[2.0]);
        *g.state_mut(close) = arr(&[10.0]);
        g.stabilize(&mut pool, &Instant::MIN);
        let o = g.view(out).as_slice().unwrap().to_vec();
        assert!(
            (o[0] - 2.0).abs() < 1e-12,
            "reinvested holdings {} != 2.0",
            o[0]
        );
    }
}
